MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -265,74 +265,290 @@ SetUefiImageMemoryAttributes (
+    }
+
+    RemoveEntryList (ArrayOfListEntriesToBeMerged[ListToBeMergedIndex]);
+    OrderedInsertImageRecordListEntry (ImagePropertiesRecordList, ArrayOfListEntriesToBeMerged[ListToBeMergedIndex]);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Merges every LIST_ENTRY within ArrayOfListEntriesToBeMerged into ImagePropertiesRecordList
+
+  @param[in]  ListToMergeInto                 Pointer to the head of a list of IMAGE_PROPERTIES_RECORD entries
+                                              into which the input ListToBeMerged will be merged
+  @param[in]  ListToBeMerged                  Pointer to the head of a list of IMAGE_PROPERTIES_RECORD entries
+                                              which will be merged into ListToMergeInto
+  @param[in]  ListToBeMergedCount             Number of IMAGE_PROPERTIES_RECORD entries in ListToBeMerged
+  @param[out] ArrayOfListEntriesToBeMerged    Pointer to an allocated array of LIST_ENTRY* which were merged
+                                              into the input ListToMergeInto. This array should be size
+                                              ListToBeMergedCount * sizeof(LIST_ENTRY*)
+
+  @retval EFI_SUCCESS                         ArrayOfListEntriesToBeMerged was successfully merged into
+                                              ImagePropertiesRecordList
+  @retval EFI_OUT_OF_RESOURCES                Failed to allocate memory
+  @retval EFI_INVALID_PARAMETER               ListToMergeInto was NULL                  OR
+                                              ListToBeMerged was NULL                   OR
+                                              ArrayOfListEntriesToBeMerged was NULL     OR
+                                              ListToBeMergedCount was zero
+  @retval other                               Return value of OrderedInsertImageRecordListEntry()
+**/
+STATIC
+EFI_STATUS
+MergeImagePropertiesRecordLists (
+  IN  LIST_ENTRY  *ListToMergeInto,
+  IN  LIST_ENTRY  *ListToBeMerged,
+  IN  UINTN       ListToBeMergedCount,
+  OUT LIST_ENTRY  **ArrayOfMergedElements
+  )
+{
+  UINTN       Index = 0;
+  EFI_STATUS  Status;
+
+  if ((ListToMergeInto == NULL) || (ListToBeMerged == NULL) ||
+      (ArrayOfMergedElements == NULL) || (ListToBeMergedCount == 0))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Insert each entry in the list to be merged into the
+  while (!IsListEmpty (ListToBeMerged) && Index < ListToBeMergedCount) {
+    ArrayOfMergedElements[Index] = ListToBeMerged->ForwardLink;
+    RemoveEntryList (ArrayOfMergedElements[Index]);
+    Status = OrderedInsertImageRecordListEntry (ListToMergeInto, ArrayOfMergedElements[Index++]);
+    if (EFI_ERROR (Status)) {
+      break;
+    }
+  }
+
+  // If we did not merge all elements of the list, unmerge them and free the input array
+  if (!IsListEmpty (ListToBeMerged)) {
+    OrderedInsertImagePropertiesRecordArray (ListToBeMerged, ArrayOfMergedElements, Index - 1);
+    FreePool (*ArrayOfMergedElements);

```

Comment:
Do we want to free ArrayOfMergedElements in the failure case? It seems odd to me to free a caller allocated buffer in this case. I.e. from the caller perspective, I would assume I allocate my buffer, pass it in, check status, if status failed, free all my memory I allocated.

Reply:
I agree, but this PR should be scoped to just the special range updates (the only reason this is showing up as a diff is because I needed to move this logic to avoid forward declaration). I'll create a PR after this goes in to change the interface.